src/lib.rs

```diff
@@ -32,44 +32,69 @@ lalrpop_mod!(#[allow(clippy::all)] pub parser);
-    let mut policies: Vec<PolicyFile> = Vec::new();
-    let mut errors = CascadeErrors::new();
+/// In order to convert the compiled CIL policy into a usable policy, you must use secilc.
+pub fn compile_system_policies(
+    input_files: Vec<&str>,
+    system_names: Vec<String>,
+) -> Result<HashMap<String, String>, error::CascadeErrors> {
+    let policies = get_policies(input_files)?;
+    compile_system_policies_internal(policies, system_names, false)
+}
 
-    for f in input_files {
-        let policy_str = match std::fs::read_to_string(&f) {
-            Ok(s) => s,
-            Err(e) => {
-                errors.add_error(e);
-                continue;
-            }
-        };
-        let p = match parse_policy(&policy_str) {
-            Ok(p) => p,
-            Err(evec) => {
-                for e in evec {
-                    // TODO: avoid String duplication
-                    errors.add_error(error::ParseError::new(e, f.into(), policy_str.clone()));
-                }
-                continue;
+/// Compile all of the system policies
+/// The list of input files list should contain filenames of files containing policy to be
+/// compiled.
+/// Returns a Result containing either a string of CIL policy which is the compiled result or a
+/// list of errors.
+/// In order to convert the compiled CIL policy into a usable policy, you must use secilc.
+pub fn compile_system_policies_all(
+    input_files: Vec<&str>,
+) -> Result<HashMap<String, String>, error::CascadeErrors> {
+    let mut system_names = Vec::new();
+    let policies = get_policies(input_files)?;
+    for p in &policies {
+        for e in &p.policy.exprs {
+            if let Expression::Decl(Declaration::System(s)) = e {
+                system_names.push(s.name.to_string());
             }
-        };
-
-        policies.push(PolicyFile::new(*p, SimpleFile::new(f.into(), policy_str)));
+        }
     }
-    // Stops if something went wrong for this major step.
-    errors = errors.into_result_self()?;
+    compile_system_policies_internal(policies, system_names, false)
+}
+
+/// Compile all systems into a single policy
+/// The list of input files list should contain filenames of files containing policy to be
+/// compiled.
+/// The system name is the name of the output file (out.cil by default).
+/// Returns a Result containing either a string of CIL policy which is the compiled result or a
+/// list of errors.
+/// In order to convert the compiled CIL policy into a usable policy, you must use secilc.
+pub fn compile_combined(
+    input_files: Vec<&str>,
+    system_name: String,
+) -> Result<String, error::CascadeErrors> {
+    let errors = CascadeErrors::new();
+    let policies = get_policies(input_files)?;
+    let res = compile_system_policies_internal(policies, vec![system_name.clone()], true)?;
+    let ret = res.get(&system_name);
+    errors.into_result(ret.unwrap().to_string())
```

Comment:
Okay, so ret could be `None` if `compile_system_policies_internal()` returned `Ok(_)`, but didn't store a system named `system_name` in the HashMap.  That would presumably be a bug in compile_system_policies_internal().  I think in the event of that sort of bug we have typically (for better or worse) been returning an InternalError rather than panicing.

Reply:
Also, the `to_string()` is here because HashMap get returns an `Option<&T>`, right?  So ret.unwrap() returns type `&String`?  In that case, I think that just dereferencing with * seems cleaner?  (It looks like to_string() on &String calls to_owned(), which may clone() under the hood, so `*` would avoid the clone).  All this is of course moot if I'm wrong about the types and this is actually a &str for some HashMap reason.

Reply:
I tried out using * instead of to_string(), but it doesn't seem like dereferencing a &String is allowed.

Reply:
I played with this a little on my system and this change compiles and passes tests for me:

```
-    let res = compile_system_policies_internal(policies, vec![system_name.clone()], true)?;
-    let ret = res.get(&system_name);
-    errors.into_result(ret.unwrap().to_string())
+    let mut res = compile_system_policies_internal(policies, vec![system_name.clone()], true)?;
+    let ret = res.remove(&system_name);
+    errors.into_result(ret.unwrap())
```

I believe the issue that you're hitting is that the code actually needed the implicit clone() to work from an ownership perspective.  Since res is a `HashMap<String, String>`, ret is a reference into that HashMap.  Trying to take ownership of that reference by dereferencing isn't legal because the HashMap still owns it, so deref() tries to implicitly copy(), which fails because String doesn't impliment the Copy trait.

However, we don't need to retain ownership of the HashMap.  What we really want is to take ownership of the string we'll be returning from the HashMap, and then transfer that ownership to the caller.  Switching the get() call to a remove() call accomplishes that.